theseus/embodied/robot/forward_kinematics.py

```diff
@@ -0,0 +1,146 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import torch
+
+from typing import List, Optional
+
+from theseus.labs.lie_functional import se3
+from .robot import Robot
+from .link import Link
+from .joint import Joint
+
+
+# TODO: Add support for joints with DOF>1
+
+
+def ForwardKinematicsFactory(robot: Robot, link_names: Optional[List[str]] = None):
+    links: List[Link] = []
+
+    if link_names is None:
+        links = robot.links
+    else:
+        links = [robot.link_map[name] for name in link_names]
+
+    link_ids: List[int] = [link.id for link in links]
+
+    ancestors: List[Link] = []
+    joint_ids: List[int] = []
+    for link in links:
+        ancestors += [anc for anc in link.ancestors]
+        joint_ids += link.angle_ids
+    pose_ids = sorted(list(set([anc.id for anc in ancestors] + link_ids)))
+    joint_ids = sorted(list(set(joint_ids)))
```

Comment:
Same comment as https://github.com/facebookresearch/theseus/pull/417#discussion_r1091528884

Reply:
- Yeah, I will refactor the code with `link.joint` and `link.parent`.
- I think keeping the ancestor ids is important in lots of applications. This makes it possible to get the sparsity structure when the robot model is used in collision.
- We have `robot.joints` and the users get the joint names by `robot.joints[joint_id].name`.